### ISSUE TITLE
Question

### ISSUE THREAD

**Issue by Mellnik** [0xbigshaq/luajit-pwn#1]
Just found this by googling and wanted to ask if this exploit works when loading as text file rather than bytecode?

**Comment by 0xbigshaq** [comment]
Hi, not exactly
The exploit relies on the [exp-gen.py](https://github.com/0xbigshaq/luajit-pwn/blob/main/sandbox-escape/exp-gen.py#L18-L19) file, which cooks a very powerful primitive that is leveraged later-on in the Lua code. It is done by hot-patching couple of Lua bytecode instructions(More about it can be found [here](https://0xbigshaq.github.io/2022/12/30/luajit-sandbox-escape/)).

However, When loading a script, the interpreter has an "_auto-detect_" feature: It checks whether the very first bytes of the script are `LUA_SIGNATURE` magic bytes -> if it is, it will call `bc_read` to read the bytecode([#1](https://github.com/LuaJIT/LuaJIT/blob/v2.1.0-beta3/src/lj_load.c#L41)). Some vendors that embeds Lua within their products sometimes forget to disable this _hacky_ feature([#2](https://github.com/redis/redis/commit/fdf9d455098f54f7666c702ae464e6ea21e25411), [#3](https://benmmurphy.github.io/blog/2015/06/04/redis-eval-lua-sandbox-escape/)) so you might get lucky.

In other words, even if you see that the target program normally accepts text-based input: try to feed it with a specially-crafted bytecode & hope for a segfault :^)
And if you get a generic error: you probably going to need a different exploit 🫠 

I hope it answers your question/was informative enough

**Comment by Mellnik** [comment]
I wanted to add that loading bytecode is never considered secure anyway. And when you really load untrusted code you set the mode parameter to `t` anyway. So in the end there is nothing wrong with LuaJIT but rather users who misuse LuaJIT.

**Comment by 0xbigshaq** [comment]
>_I wanted to add that loading bytecode is never considered secure anyway._

uhm yes, that part is obvious. the luajit faq docs says that as well.

>_And when you really load untrusted code you set the mode parameter to `t` anyway_

that's only valid for `loadfile()` and friends. Not the interpreter binary itself(afaik there's no `--no-bytecode` flag when running `luajit file.lua`). Plus, the issue lays in the fact that this dangerous feature is implicitly enabled while it should be explicit. The default `m(ode)` value should be set to interpret text-based Lua files only, and if you wish to add bytecode support then set it to be `m='b'`(ytecode) explicitly, or whatever the flag value is. It could prevent a lot of people from shooting themselves in the foot.

> _So in the end there is nothing wrong with LuaJIT_

That's debatable.

I'm going to lock this thread since you already got the answer to your initial question and this is github, not a discord groupchat(no offense)
